### PR TITLE
Implement Saved Commands

### PR DESCRIPTION
**Pull request by Abyss777**
Saved Command edit dialog

image

Linking as usual

image

Here is Send Command dialog

image

User can select "New ..." if allowed or supported Saved Commands

There is a preview of Saved Command parameters

image

Or new command can be set

image

Another thing was changed here is that we have a lot of the same permission windows, removed them, now instead of `DeviceXxx`, `GroupXxx`, `UserXxx` just `Xxx`
````
'Traccar.view.permissions.Devices',
'Traccar.view.permissions.Groups',
'Traccar.view.permissions.Geofences',
'Traccar.view.permissions.Calendars',
'Traccar.view.permissions.Users',
'Traccar.view.permissions.ComputedAttributes',
'Traccar.view.permissions.Drivers',
'Traccar.view.permissions.SavedCommands'
```` 
More comments in code

### CODE REVIEW

File: web/app/model/Command.js [hunk 1]
```diff
@@ -20,12 +20,26 @@ Ext.define('Traccar.model.Command', {
...
+        name: 'description',
+        type: 'string'
+    }, {
         name: 'deviceId',
         type: 'int'
     }, {
         name: 'type',
         type: 'string'
     }, {
+        name: 'name',
```
Comment (Abyss777) [inline web/app/model/Command.js]: Added not persist field to translate `type`

File: web/app/store/AllCommandTypes.js [hunk 2]
```diff
@@ -15,23 +16,17 @@
  * along with this program. If not, see <http://www.gnu.org/licenses/>.
  */
 
-Ext.define('Traccar.view.permissions.DeviceGeofences', {
-    extend: 'Traccar.view.permissions.Base',
-    xtype: 'deviceGeofencesView',
+Ext.define('Traccar.store.AllCommandTypes', {
```
Comment (Abyss777) [inline web/app/store/AllCommandTypes.js]: Store to all possible command types

File: web/app/store/DeviceCommands.js [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program. If not, see <http://www.gnu.org/licenses/>.
+ */
+
+Ext.define('Traccar.store.DeviceCommands', {
```
Comment (Abyss777) [inline web/app/store/DeviceCommands.js]: Store for currently supported commands by device

File: web/app/view/dialog/SavedCommandController.js [hunk 4]
```diff
@@ -15,25 +15,30 @@
...
+    extend: 'Traccar.view.dialog.BaseEditController',
+    alias: 'controller.savedCommand',
 
     defaultFieldConfig: {
         allowBlank: false
     },
 
-    onSelect: function (selected) {
-        var i, config, command, parameters, parameter;
+    onTypeChange: function (combo, newValue) {
```
Comment (Abyss777) [inline web/app/view/dialog/SavedCommandController.js]: React on changing type in any way, not only by user selecting.

File: web/app/view/dialog/SavedCommandController.js [hunk 5]
```diff
@@ -15,25 +15,30 @@
...
-        command = Ext.getStore('KnownCommands').getById(selected.getValue());
+        command = Ext.getStore('KnownCommands').getById(newValue);
         if (command && command.get('parameters')) {
             parameters = command.get('parameters');
             for (i = 0; i < parameters.length; i++) {
                 parameter = new Traccar.model.KnownAttribute(parameters[i]);
                 config = Ext.clone(this.defaultFieldConfig);
                 config.key = parameter.get('key');
                 config.fieldLabel = parameter.get('name');
+                if (record.get('attributes')) {
```
Comment (Abyss777) [inline web/app/view/dialog/SavedCommandController.js]: Initialize values if we are editing Saved Command

File: web/app/view/dialog/SendCommandController.js [hunk 6]
```diff
@@ -0,0 +1,93 @@
...
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program. If not, see <http://www.gnu.org/licenses/>.
+ */
+
+Ext.define('Traccar.view.dialog.SendCommandController', {
+    extend: 'Traccar.view.dialog.SavedCommandController',
```
Comment (Abyss777) [inline web/app/view/dialog/SendCommandController.js]: Reuse `SavedCommandContoller`

File: web/app/view/dialog/SendCommandController.js [hunk 7]
```diff
@@ -0,0 +1,93 @@
...
+    onSendClick: function (button) {
+        var record;
+        this.fillAttributes(button);
+        record = button.up('window').down('form').getRecord();
+
+        Ext.Ajax.request({
+            scope: this,
+            url: 'api/commands/send',
+            jsonData: record.getData({
+                persist: true
```
Comment (Abyss777) [inline web/app/view/dialog/SendCommandController.js]: Opposite meaning: "do not return fields with `persist: false`"

File: web/app/view/dialog/SendCommandController.js [hunk 8]
```diff
@@ -0,0 +1,93 @@
...
+    },
+
+    onCommandSelect: function (selected) {
+        var record, form, command = selected.getStore().getById(selected.getValue());
+        command.set('deviceId', this.getView().deviceId);
+        form = selected.up('window').down('form');
+        record = form.getRecord();
+        form.loadRecord(command);
+        if (record && command.get('type') === record.get('type')) {
+            this.onTypeChange(this.lookupReference('commandType'), command.get('type'));
```
Comment (Abyss777) [inline web/app/view/dialog/SendCommandController.js]: Emulate type changing in case we selected another Saved Command with same type, but other parameters.

File: web/app/view/edit/Devices.js [hunk 9]
```diff
@@ -37,67 +37,54 @@ Ext.define('Traccar.view.edit.Devices', {
 
     tbar: {
         componentCls: 'toolbar-header-style',
+        defaults: {
```
Comment (Abyss777) [inline web/app/view/edit/Devices.js]: Optimized common fields

File: web/app/view/edit/DevicesController.js [hunk 10]
```diff
@@ -114,22 +115,31 @@ Ext.define('Traccar.view.edit.DevicesController', {
...
-        typesStore = dialog.lookupReference('commandType').getStore();
-        typesStore.getProxy().setExtraParam('deviceId', deviceId);
+        dialog = Ext.create('Traccar.view.dialog.SendCommand');
+        dialog.deviceId = deviceId;
+        dialog.online = online;
+
+        commandsStore = dialog.lookupReference('commandsComboBox').getStore();
+        commandsStore.getProxy().setExtraParam('deviceId', deviceId);
+        commandsStore.removeAll();
+        if (!Traccar.app.getPreference('limitCommands', false)) {
```
Comment (Abyss777) [inline web/app/view/edit/DevicesController.js]: If user can post new command we are adding empty one on first place

**Comment by tananaev** [comment]
I think in the send command dialog we should just have "Command" label for the combobox. It includes regular commands, right?

File: web/app/model/Command.js [hunk 11]
```diff
@@ -20,12 +20,26 @@ Ext.define('Traccar.model.Command', {
...
+        convert: function (v, rec) {
+            var name = Strings['command' + rec.get('type').charAt(0).toUpperCase() + rec.get('type').slice(1)];
```
Comment (tananaev) [inline web/app/model/Command.js]: We have similar code in other places. Should be have some helper method to do it?

File: web/l10n/en.json [hunk 12]
```diff
@@ -69,6 +69,9 @@
     "sharedTypeBoolean": "Boolean",
     "sharedTimezone": "Timezone",
     "sharedInfoTitle": "Info",
+    "sharedSavedCommand": "Saved Command",
+    "sharedSavedCommands": "Saved Commands",
+    "sharedNew": "New ...",
```
Comment (tananaev) [inline web/l10n/en.json]: Use ellipsis character instead of 3 dots.

File: web/app/view/edit/DevicesController.js [hunk 13]
```diff
@@ -114,22 +115,31 @@ Ext.define('Traccar.view.edit.DevicesController', {
...
-
-        typesStore = dialog.lookupReference('commandType').getStore();
-        typesStore.getProxy().setExtraParam('deviceId', deviceId);
+        dialog = Ext.create('Traccar.view.dialog.SendCommand');
+        dialog.deviceId = deviceId;
+        dialog.online = online;
+
+        commandsStore = dialog.lookupReference('commandsComboBox').getStore();
+        commandsStore.getProxy().setExtraParam('deviceId', deviceId);
+        commandsStore.removeAll();
```
Comment (tananaev) [inline web/app/view/edit/DevicesController.js]: Maybe we should clear it when we leave the dialog?

**Comment by Abyss777** [comment]
Done changes

image

Also used renderer instead of not-persist field in Saved Command

**Comment by tananaev** [comment]
Merged, thanks.
